File: scripts/reconstruct.py
import argparse
import itertools
from pathlib import Path

import cv2
import numpy as np

from sfm.core.features import detect, match
from sfm.core.geometry import (
    angle_between_rotations,
    camera_center,
    relative_pose,
)
from sfm.core.pipeline import Session
from sfm.dataset import duplicate_views, load_middlebury

DATA_DIR = Path("data/templeSparseRing")
PAR_FILE = DATA_DIR / "templeSR_par.txt"
MIN_PAIR_MATCHES = 30


def build_session(cameras, seed):
    detector = cv2.SIFT_create()
    matcher = cv2.BFMatcher()
    features = {}
    for i, camera in enumerate(cameras):
        image = cv2.imread(str(DATA_DIR / camera.name), cv2.IMREAD_GRAYSCALE)
        features[i] = detect(image, detector)
    matches = {}
    for i, j in itertools.combinations(range(len(cameras)), 2):
        pairs = match(features[i], features[j], matcher)
        if len(pairs) >= MIN_PAIR_MATCHES:
            matches[(i, j)] = pairs
    return Session(features, matches, cameras[0].k, seed), features, matches


def report(result, cameras, anchor):
    recon = result.reconstruction
    scale = None
    print(f"\n{len(recon.poses)} cameras, {len(recon.points)} points, "
          f"rms {result.report.rms:.4f} px, null space {result.report.null_space_dimension}")
    print("camera | angle error | predicted | center error | predicted")
    for c in sorted(recon.poses):
        r, t = recon.poses[c]
        r_true, t_true = relative_pose(cameras[anchor].r, cameras[anchor].t,
                                       cameras[c].r, cameras[c].t)
        if scale is None and np.linalg.norm(t_true) > 1e-9:
            scale = np.linalg.norm(t_true) / max(np.linalg.norm(t), 1e-12)
        angle = angle_between_rotations(r, r_true)
        center = np.linalg.norm(camera_center(r, t) * (scale or 1.0)
                                - camera_center(r_true, t_true))
        print(f"{c + 1:6d} | {angle:11.4f} | {result.report.angle_sigma.get(c, 0.0):9.4f} | "
              f"{center:12.6f} | {result.report.center_sigma.get(c, 0.0) * (scale or 1.0):9.6f}")
    for item in result.rejected:
        print(f"  rejected {item.camera + 1}: {item.reason} "
              f"(matches {item.matches}, inliers {item.inliers}"
              + (f", excess null {item.excess_null_dimensions}"
                 if item.excess_null_dimensions else "") + ")")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--seed-pair", type=int, nargs=2, default=[12, 13])
    parser.add_argument("--seed", type=int, default=7)
    args = parser.parse_args()

    cameras = load_middlebury(PAR_FILE)
    for i, j in duplicate_views(cameras):
        print(f"warning: views {i + 1} and {j + 1} share the same centre")

    session, _, _ = build_session(cameras, args.seed)
    left, right = args.seed_pair[0] - 1, args.seed_pair[1] - 1
    result = session.run((left, right))
    report(result, cameras, left)


if __name__ == "__main__":
    main()

File: web/scripts/reconstruct.mjs
import { readFileSync } from 'node:fs';

import {
  MIN_PAIR_MATCHES,
  Session,
  detectImage,
  matchFeatures,
} from '../src/pipeline.js';
import {
  angleBetweenRotations,
  cameraCenter,
  matmul,
  norm,
  relativePose,
  subtract,
  transpose,
} from '../src/geometry.js';

const DATASET = './tests/temple.json';
const SEED_PAIR = [11, 12];

function loadDataset() {
  const data = JSON.parse(readFileSync(DATASET, 'utf8'));
  const images = data.images.map((pixels) => ({
    width: data.width,
    height: data.height,
    data: Float32Array.from(pixels),
  }));
  return { cameras: data.cameras, images };
}

function buildSession({ cameras, images }) {
  const started = Date.now();
  const features = new Map();
  images.forEach((image, index) => {
    features.set(index, detectImage(image));
  });
  const detectMs = Date.now() - started;

  const matchStart = Date.now();
  const matches = new Map();
  for (let i = 0; i < images.length; i += 1) {
    for (let j = i + 1; j < images.length; j += 1) {
      const pairs = matchFeatures(features.get(i), features.get(j));
      if (pairs.length >= MIN_PAIR_MATCHES) matches.set(`${i}:${j}`, pairs);
    }
  }
  const matchMs = Date.now() - matchStart;

  const counts = [...features.values()].map((list) => list.length);
  console.log(`detected ${Math.min(...counts)}-${Math.max(...counts)} features per image`
    + ` in ${(detectMs / 1000).toFixed(1)} s`);
  console.log(`matched ${matches.size} of ${(images.length * (images.length - 1)) / 2} pairs`
    + ` in ${(matchMs / 1000).toFixed(1)} s`);
  return new Session(features, matches, cameras[0].k);
}

function report(result, cameras, anchor) {
  const { reconstruction, report: summary } = result;
  console.log(`\n${reconstruction.poses.size} cameras, ${reconstruction.points.length} points,`
    + ` rms ${summary.rms.toFixed(4)} px, null space ${summary.nullSpaceDimension}`);
  console.log('camera | angle error | predicted | centre error');

  let scale = null;
  for (const camera of [...reconstruction.poses.keys()].sort((a, b) => a - b)) {
    const pose = reconstruction.poses.get(camera);
    const truth = relativePose(cameras[anchor].r, cameras[anchor].t,
      cameras[camera].r, cameras[camera].t);
    if (scale === null && norm(truth.t) > 1e-9) {
      scale = norm(truth.t) / Math.max(norm(pose.t), 1e-12);
    }
    const angle = angleBetweenRotations(pose.r, truth.r);
    const estimated = cameraCenter(pose.r, pose.t).map((v) => v * (scale ?? 1));
    const centre = norm(subtract(estimated, cameraCenter(truth.r, truth.t)));
    const predicted = summary.angleSigma.get(camera) ?? 0;
    console.log(`${String(camera + 1).padStart(6)} | ${angle.toFixed(4).padStart(11)}`
      + ` | ${predicted.toFixed(4).padStart(9)} | ${centre.toFixed(6).padStart(12)}`);
  }
  for (const item of result.rejected) {
    const extra = item.excessNullDimensions ? `, excess null ${item.excessNullDimensions}` : '';
    console.log(`  rejected ${item.camera + 1}: ${item.reason}`
      + ` (matches ${item.matches ?? 0}, inliers ${item.inliers ?? 0}${extra})`);
  }
}

const dataset = loadDataset();
const session = buildSession(dataset);
const started = Date.now();
const result = session.run(SEED_PAIR);
console.log(`reconstructed in ${((Date.now() - started) / 1000).toFixed(1)} s`);
report(result, dataset.cameras, SEED_PAIR[0]);